<a href="https://colab.research.google.com/github/msantoshidevi/hotel_booking_project/blob/hotel_booking_analyses_modification/Hotel_booking_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project Name**    - Hotel Booking Analysis



##### **Project Type**    - EDA
##### **Contribution**    - Individual
##### **Team Member**     - Santoshi Devi M

# **Project Summary -**

This exploratory data analysis(EDA) project focuses on analyzing the Hotel Booking dataset. This hotel booking dataset contain booking information for city hotel and resort hotel and includes information such as when the booking was made, length of stay, the number of adults, children and/or babies, and the number of available parking spaces, among other things.

It consists of 1,19,390 observations and 31 variables. Each observation represents a hotel booking. The dataset comprehense bookings arrive between the 1st of July of 2015 and 31st of August 2017, including bookings that effectively arrived and bookings that were cancelled.

Our analysis can help us understand about the given dataset and draw some useful insights to make some crucial business decisions regarding the factors affecting the bookings of the hotels. In hotel industry, cancellations and average daily rate are two important factors that effects the business.

By understanding the factors that are determining the cancellation of certain bookings, the hotels can take necessary precautions to reduce cancellation rate.

By understanding the patterns in average daily rate against different variables the hotels can be prepared in advance to generate more revenue and help to make more profitable business.

Our goal here is to understand such factors in the given data by performing exploratory data analysis. Exploratory Data Analysis refers to the critical process of performing initial investigations on data, so as to discover patterns, to spot anomalies to test hypothesis and to check assumptions with the help of summary statistics and graphical representations

The following are the various steps performed as a part of Exploratory Data Analysis:
* Data preparation
* Data cleaning
* Univariate analysis
* Bivariate analysis
* Multivariate analysis

**Data preparation:** Firstly, we imported libraries and dataset, some of the libraries used are NumPy, Pandas, matplotlib, seaborn. Once dataset and libraries are imported, process of analysis begins. But data have to be translated in an appropriate form. This process is known as data preparation.

**Data Cleaning:** The raw data received in the data received in the data set might not be directly suitable for analysis due to presence of unwanted data like, duplicate values, null values, outliers etc. We need to handle them first before we proceed with further analysis.

The duplicate values in the dataset to be removed, so that we get accurate patterns and numbers regarding the analysis.

The missing information for some or all features in some records, we need to either remove them or find alternatives to fill up the null values.

**Univariate Analysis:** Choose a single feature from the data and try to determine what the output or target value is, i.e. one features/variable at a time.
* Understand the trends and patterns of data.
* Analyze the frequency and other such characteristics of data.
* Know the distribution of the variables in the data.
* Visualise the relationship that may exist between different variables.

**Bivariate Analysis:** We try to analyze two features instead of one, and finally determine the classification of output we are looking for. It is a methodical statistical technique applied to a pair of variables of data to determine the empirical relationship between them. In  order words, it is meant to determine any concurrent relations.

**Multivariate Analysis:** We analyze three or more different features at a time, to understand the relationship between all the features involved.

After completing all the above processes, we have observed final conclusion from the patterns of the analysis and identified anomalies that can help increase the business objective. Based on findings, we recommand the hotels to implement necessary steps and plans to improve the marketing strategy.

# **GitHub Link -**

https://github.com/msantoshidevi/EDA_hotel_booking_project.git

# **Problem Statement**


We are here to explore a hotel booking dataset to discover important factors that govern the bookings, which contain booking information for a city hotel and resort hotel. We will analyze some important aspects of hotel bookings which will helps us identify major loopholes and give us insights which will be helpful to run profitable hotel business are as follows:
* Optimal length of stay to get the best daily rate?
* Which agent makes most number of bookings?
* which month results in high revenue?
* To predict whether or not a hotel was likely to receive a disproportionately high number of special requests?

#### **Define Your Business Objective?**

In Hotel industry, cancellation and Average Daily Rate  are two important factors that helps to run the business effectively.
- By understanding the factors that are determining the cancellation of a certain bookings, the hotels can take necessary precautions to reduce cancellation rate.
- By understanding the patterns in Average Daily Rate against different variables, the hotels can be prepared in advance to generate more revenue and help to make a profitable business.

Our goal here is to understand such factors in the given dataset by performing Exploratory Data Analysis.

# **General Guidelines** : -  

1.   Well-structured, formatted, and commented code is required.
2.   Exception Handling, Production Grade Code & Deployment Ready Code will be a plus. Those students will be awarded some additional credits.
     
     The additional credits will have advantages over other students during Star Student selection.
       
             [ Note: - Deployment Ready Code is defined as, the whole .ipynb notebook should be executable in one go
                       without a single error logged. ]

3.   Each and every logic should have proper comments.
4. You may add as many number of charts you want. Make Sure for each and every chart the following format should be answered.
        

```
# Chart visualization code
```
            

*   Why did you pick the specific chart?
*   What is/are the insight(s) found from the chart?
* Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

5. You have to create at least 20 logical & meaningful charts having important insights.


[ Hints : - Do the Vizualization in  a structured way while following "UBM" Rule.

U - Univariate Analysis,

B - Bivariate Analysis (Numerical - Categorical, Numerical - Numerical, Categorical - Categorical)

M - Multivariate Analysis
 ]





# ***Let's Begin !***

## ***1. Know Your Data***

### Import Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

### Dataset Loading

In [ ]:
# Load Dataset

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
hotel_df = pd.read_csv("/content/drive/MyDrive/Project/Hotel Bookings.csv")
hotel_df

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

### Dataset First View

In [4]:
# Dataset First Look
hotel_df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

### Dataset Rows & Columns count

In [5]:
# Dataset Rows & Columns count
hotel_df.shape

(119390, 32)

**OBSERVATION**

* There are 119390 rows and 32 columns which means there are 119390 observations and 31 variables.



### Dataset Information

In [6]:
# Dataset Info
hotel_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [7]:
# Creating a copy of dataframe
df = hotel_df.copy()

#### Duplicate Values

In [8]:
# Dataset Duplicate Value Count
df[hotel_df.duplicated()].shape

(31994, 32)

**OBSERVATION**
*   The number of duplicate rows are 31994 out of 119390.

In [9]:
df.drop_duplicates(inplace = True)

In [10]:
df.shape

(87396, 32)

**OBSERVATION**
*   After removing 31994 duplicate rows, we have 87396 rows and 32 columns in the dataset.

#### Missing Values/Null Values

In [11]:
# Missing Values/Null Values Count
df.isnull().sum().sort_values(ascending = False)[:6]

company               82137
agent                 12193
country                 452
children                  4
reserved_room_type        0
assigned_room_type        0
dtype: int64

In [12]:
df['agent'].unique()

array([ nan, 304., 240., 303.,  15., 241.,   8., 250., 115.,   5., 175.,
       134., 156., 243., 242.,   3., 105.,  40., 147., 306., 184.,  96.,
         2., 127.,  95., 146.,   9., 177.,   6., 143., 244., 149., 167.,
       300., 171., 305.,  67., 196., 152., 142., 261., 104.,  36.,  26.,
        29., 258., 110.,  71., 181.,  88., 251., 275.,  69., 248., 208.,
       256., 314., 126., 281., 273., 253., 185., 330., 334., 328., 326.,
       321., 324., 313.,  38., 155.,  68., 335., 308., 332.,  94., 348.,
       310., 339., 375.,  66., 327., 387., 298.,  91., 245., 385., 257.,
       393., 168., 405., 249., 315.,  75., 128., 307.,  11., 436.,   1.,
       201., 183., 223., 368., 336., 291., 464., 411., 481.,  10., 154.,
       468., 410., 390., 440., 495., 492., 493., 434.,  57., 531., 420.,
       483., 526., 472., 429.,  16., 446.,  34.,  78., 139., 252., 270.,
        47., 114., 301., 193., 182., 135., 350., 195., 352., 355., 159.,
       363., 384., 360., 331., 367.,  64., 406., 16

In [13]:
df['company'].unique()

array([ nan, 110., 113., 270., 178., 240., 154., 144., 307., 268.,  59.,
       204., 312., 318.,  94., 174., 274., 195., 223., 317., 281., 118.,
        53., 286.,  12.,  47., 324., 342., 373., 371., 383.,  86.,  82.,
       218.,  88.,  31., 397., 392., 405., 331., 367.,  20.,  83., 416.,
        51., 395., 102.,  34.,  84., 360., 394., 457., 382., 461., 478.,
       386., 112., 486., 421.,   9., 308., 135., 224., 504., 269., 356.,
       498., 390., 513., 203., 263., 477., 521., 169., 515., 445., 337.,
       251., 428., 292., 388., 130., 250., 355., 254., 543., 531., 528.,
        62., 120.,  42.,  81., 116., 530., 103.,  39.,  16.,  92.,  61.,
       501., 165., 291., 290.,  43., 325., 192., 108., 200., 465., 287.,
       297., 490., 482., 207., 282., 437., 225., 329., 272.,  28.,  77.,
       338.,  72., 246., 319., 146., 159., 380., 323., 511., 407., 278.,
        80., 403., 399.,  14., 137., 343., 346., 347., 349., 289., 351.,
       353.,  54.,  99., 358., 361., 362., 366., 37

**OBSERVATION**

*   Since, company and agent columns have company number and agent numbers as
data.
*   There may be some cases when customer didnt booked hotel via any agent or via any company.
*    So in that case values can be null under these columns.
*   We will replace null values by 0 in these columns

In [14]:
df[['company','agent']] = df[['company','agent']].fillna(0)

In [15]:
df['children'].unique()

array([ 0.,  1.,  2., 10.,  3., nan])

**OBSERVATION**

*  This column 'children' has 0 as value which means 0 children were present in group of customers who made that transaction.
*   So, 'nan' values are the missing values due to error of recording data.

In [16]:
df['children'].fillna(df['children'].mean(), inplace = True)

In [17]:
df['country'].unique()

array(['PRT', 'GBR', 'USA', 'ESP', 'IRL', 'FRA', nan, 'ROU', 'NOR', 'OMN',
       'ARG', 'POL', 'DEU', 'BEL', 'CHE', 'CN', 'GRC', 'ITA', 'NLD',
       'DNK', 'RUS', 'SWE', 'AUS', 'EST', 'CZE', 'BRA', 'FIN', 'MOZ',
       'BWA', 'LUX', 'SVN', 'ALB', 'IND', 'CHN', 'MEX', 'MAR', 'UKR',
       'SMR', 'LVA', 'PRI', 'SRB', 'CHL', 'AUT', 'BLR', 'LTU', 'TUR',
       'ZAF', 'AGO', 'ISR', 'CYM', 'ZMB', 'CPV', 'ZWE', 'DZA', 'KOR',
       'CRI', 'HUN', 'ARE', 'TUN', 'JAM', 'HRV', 'HKG', 'IRN', 'GEO',
       'AND', 'GIB', 'URY', 'JEY', 'CAF', 'CYP', 'COL', 'GGY', 'KWT',
       'NGA', 'MDV', 'VEN', 'SVK', 'FJI', 'KAZ', 'PAK', 'IDN', 'LBN',
       'PHL', 'SEN', 'SYC', 'AZE', 'BHR', 'NZL', 'THA', 'DOM', 'MKD',
       'MYS', 'ARM', 'JPN', 'LKA', 'CUB', 'CMR', 'BIH', 'MUS', 'COM',
       'SUR', 'UGA', 'BGR', 'CIV', 'JOR', 'SYR', 'SGP', 'BDI', 'SAU',
       'VNM', 'PLW', 'QAT', 'EGY', 'PER', 'MLT', 'MWI', 'ECU', 'MDG',
       'ISL', 'UZB', 'NPL', 'BHS', 'MAC', 'TGO', 'TWN', 'DJI', 'STP',
       'KNA', 'E

**OBSERVATION**
*   'country' column represents the country of origin of customer.
*   Since, this column has datatype of string. We will replace the missing value with the mode of 'country' column.

In [18]:
df['country'].fillna('others', inplace = True)

In [19]:
df.isnull().sum().sort_values(ascending = False)[:6]

hotel                          0
is_canceled                    0
reservation_status             0
total_of_special_requests      0
required_car_parking_spaces    0
adr                            0
dtype: int64

In [20]:
df[(df["adults"] == 0) & (df['children'] == 0) & (df['babies'] == 0)][["adults","children","babies","is_canceled"]]

adults  children  babies  is_canceled
2224         0       0.0       0            0
2409         0       0.0       0            0
3181         0       0.0       0            0
3684         0       0.0       0            0
3708         0       0.0       0            0
...        ...       ...     ...          ...
115029       0       0.0       0            0
115091       0       0.0       0            0
116251       0       0.0       0            0
116534       0       0.0       0            0
117087       0       0.0       0            0

[166 rows x 4 columns]

**OBSERVATION**

There are some rows with total number of adults, children or babies equal to zero. As adults, babies, and children can't be zero at the same time, so dropping the rows having all these zero at the same time because will mean that the hotel/room was not even booked.

In [21]:
df[df['adults']+df['babies']+df['children'] == 0].shape

(166, 32)

In [22]:
df.drop(df[df['adults']+df['babies']+df['children'] == 0].index, inplace = True)

In [23]:
df.shape

(87230, 32)

### What did you know about your dataset?

The dataset contains information about bookings and revenue for resots and hotels properties from July 2015 to August 2017. After filling missing values, removing duplicates and unwanted data removing, dataset consists of 87230 rows and 32 columns provides comprehensive details for analysis.

Our objective to analyze the data and gain insights into the best time for hotel bookings and the optimal length of stay to obtain the best rates. Additionally, we will explore other key aspects of the dataset to uncover valuable information.

By conducting a thorough analysis of this dataset, we aim to provide meaningful insights and recommendations to improve the booking strategies and revenue optimization for resort and hotel properties.

## ***2. Understanding Your Variables***

In [24]:
# Dataset Columns
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

In [25]:
# Dataset Describe
df.describe()

is_canceled     lead_time  arrival_date_year  \
count  87230.000000  87230.000000       87230.000000   
mean       0.275238     79.971019        2016.210352   
std        0.446637     86.058683           0.686064   
min        0.000000      0.000000        2015.000000   
25%        0.000000     11.000000        2016.000000   
50%        0.000000     49.000000        2016.000000   
75%        1.000000    125.000000        2017.000000   
max        1.000000    737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count              87230.000000               87230.000000   
mean                  26.835091                  15.815832   
std                   13.669216                   8.835545   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   27.000000                  16.000000   
75%                   37.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights        adults  \
count             87230.000000          87230.000000  87230.000000   
mean                  1.004609              2.623925      1.879365   
std                   1.027408              2.039830      0.621724   
min                   0.000000              0.000000      0.000000   
25%                   0.000000              1.000000      2.000000   
50%                   1.000000              2.000000      2.000000   
75%                   2.000000              4.000000      2.000000   
max                  19.000000             50.000000     55.000000   

           children        babies  is_repeated_guest  previous_cancellations  \
count  87230.000000  87230.000000       87230.000000            87230.000000   
mean       0.138904      0.010845           0.038565                0.030402   
std        0.456264      0.113704           0.192556                0.369344   
min        0.000000      0.000000           0.000000                0.000000   
25%        0.000000      0.000000           0.000000                0.000000   
50%        0.000000      0.000000           0.000000                0.000000   
75%        0.000000      0.000000           0.000000                0.000000   
max       10.000000     10.000000           1.000000               26.000000   

       previous_bookings_not_canceled  booking_changes         agent  \
count                    87230.000000     87230.000000  87230.000000   
mean                         0.184054         0.268497     81.089258   
std                          1.733033         0.710633    109.977255   
min                          0.000000         0.000000      0.000000   
25%                          0.000000         0.000000      9.000000   
50%                          0.000000         0.000000      9.000000   
75%                          0.000000         0.000000    240.000000   
max                         72.000000        18.000000    535.000000   

            company  days_in_waiting_list           adr  \
count  87230.000000          87230.000000  87230.000000   
mean      10.984948              0.746291    106.518031   
std       53.956668             10.001001     54.891227   
min        0.000000              0.000000     -6.380000   
25%        0.000000              0.000000     72.250000   
50%        0.000000              0.000000     98.200000   
75%        0.000000              0.000000    134.100000   
max      543.000000            391.000000   5400.000000   

       required_car_parking_spaces  total_of_special_requests  
count                 87230.000000               87230.000000  
mean                      0.084306                   0.698934  
std                       0.281659                   0.832051  
min                       0.000000                   0.000000  
25%                       0.000000                   0.000000  
50%                       0

### Variables Description

**'hotel':** Name of hotel ( City or Resort) (object)

**'is_canceled':** Whether the booking is canceled or not (0 for no canceled and 1 for canceled) (int64)

**'lead_time':** time (in days) between booking transaction and actual arrival (int64)

**'arrival_date_year':** Year of arrival (int64)

**'arrival_date_month':** month of arrival (object)

**'arrival_date_week_number':** week number of arrival date (int64)

**'arrival_date_day_of_month':** Day of month of arrival date (int64)

**'stays_in_weekend_nights':** No. of weekend nights spent in a hotel (int64)

**'stays_in_week_nights':** No. of weeknights spent in a hotel (int64)

**'adults':** No. of adults in single booking record (int64)

**'children':** No. of children in single booking record. (float64)

**'babies':** No. of babies in single booking record. (int64)

**'meal':** Type of meal chosen BB – Bed & Breakfast , HB – only two meals including breakfast meal
, FB – breakfast, lunch, and dinner (object)

**'country':** Country of origin of customers (as mentioned by them) (object)

**'market_segment':** What segment via booking was made and for what purpose. - TA: Travel agents, TO: Tour operators (object)

**'distribution_channel':** Via which medium booking was made.(object)

**'is_repeated_guest':**  Whether the customer has made any booking before(0 for No and 1 for Yes) (int64)

**'previous_cancellations':** No. of previous canceled bookings. (int64)

**'previous_bookings_not_canceled':** No. of previous non-canceled bookings. (int64)

**'reserved_room_type':** Room type reserved by a customer. (object)

**'assigned_room_type':** Room type assigned to the customer. (object)

**'booking_changes':** No. of booking changes done by customers (int64)

**'deposit_type':** Type of deposit at the time of making a booking (No deposit/ Refundable/ No refund) (object)

**'agent':** Id of agent for booking (float64)

**'company':** Id of the company making a booking (float64)

**'days_in_waiting_list':** No. of days on waiting list. (int64)

**'customer_type':** Type of customer(Transient, Group, etc.) (object)

**'adr':** Average Daily rate. (float64)

**'required_car_parking_spaces':** No. of car parking asked in booking (int64)

**'total_of_special_requests':** total no. of special request. (int64)

**'reservation_status':** Whether a customer has checked out or canceled,or not showed (object)

**'reservation_status_date':** Date of making reservation status. (object)

### Check Unique Values for each variable.

In [26]:
for i in df.columns.tolist():
  print(f"No. of unique values in {i} is {df[i].nunique()}")

No. of unique values in hotel is 2
No. of unique values in is_canceled is 2
No. of unique values in lead_time is 479
No. of unique values in arrival_date_year is 3
No. of unique values in arrival_date_month is 12
No. of unique values in arrival_date_week_number is 53
No. of unique values in arrival_date_day_of_month is 31
No. of unique values in stays_in_weekend_nights is 17
No. of unique values in stays_in_week_nights is 33
No. of unique values in adults is 14
No. of unique values in children is 6
No. of unique values in babies is 5
No. of unique values in meal is 5
No. of unique values in country is 178
No. of unique values in market_segment is 8
No. of unique values in distribution_channel is 5
No. of unique values in is_repeated_guest is 2
No. of unique values in previous_cancellations is 15
No. of unique values in previous_bookings_not_canceled is 73
No. of unique values in reserved_room_type is 9
No. of unique values in assigned_room_type is 11
No. of unique values in booking_cha

In [27]:
# Check Unique Values for each variable.
df['hotel'].unique()

array(['Resort Hotel', 'City Hotel'], dtype=object)

There are 2 kinds of Hotels. one is Resort Hotel and other is City Hotel.

In [28]:
df['is_canceled'].unique()

array([0, 1])

Whether booking is canceled or not is represented in "is_canceled". It consits of 2 value 0 and 1. 0 represents not canceled and 1 represents canceled booking.

In [29]:
df['lead_time'].unique()

array([342, 737,   7,  13,  14,   0,   9,  85,  75,  23,  35,  68,  18,
        37,  12,  72, 127,  78,  48,  60,  77,  99, 118,  95,  96,  69,
        45,  40,  15,  36,  43,  70,  16, 107,  47, 113,  90,  50,  93,
        76,   3,   1,  10,   5,  17,  51,  71,  63,  62, 101,   2,  81,
       368, 364, 324,  79,  21, 109, 102,   4,  98,  92,  26,  73, 115,
        86,  52,  29,  30,  33,  32,   8, 100,  44,  80,  97,  64,  39,
        34,  27,  82,  94, 110, 111,  84,  66, 104,  28, 258, 112,  65,
        67,  55,  88,  54, 292,  83, 105, 280, 394,  24, 103, 366, 249,
        22,  91,  11, 108, 106,  31,  87,  41, 304, 117,  59,  53,  58,
       116,  42, 321,  38,  56,  49, 317,   6,  57,  19,  25, 315, 123,
        46,  89,  61, 312, 299, 130,  74, 298, 119,  20, 286, 136, 129,
       124, 327, 131, 460, 140, 114, 139, 122, 137, 126, 120, 128, 135,
       150, 143, 151, 132, 125, 157, 147, 138, 156, 164, 346, 159, 160,
       161, 333, 381, 149, 154, 297, 163, 314, 155, 323, 340, 35

In [30]:
df['arrival_date_year'].unique()

array([2015, 2016, 2017])

'arrival_date_year' variable consists of 3 values - 2015, 2016 and 2017.

In [31]:
df['arrival_date_month'].unique()

array(['July', 'August', 'September', 'October', 'November', 'December',
       'January', 'February', 'March', 'April', 'May', 'June'],
      dtype=object)

'arrival_date_month' variable consists of 12 values. Customers arrive through out the year.

In [32]:
df['arrival_date_week_number'].unique()

array([27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49, 50, 51, 52, 53,  1,  2,  3,  4,  5,  6,  7,
        8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
       25, 26])

In [33]:
df['arrival_date_day_of_month'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31])

In [34]:
df['stays_in_weekend_nights'].unique()

array([ 0,  1,  2,  4,  3,  6, 13,  8,  5,  7, 12,  9, 16, 18, 19, 10, 14])

In [35]:
df['adults'].unique()

array([ 2,  1,  3,  4, 40, 26, 50, 27, 55, 20,  6,  5, 10,  0])

In [36]:
df['children'].unique()

array([ 0.        ,  1.        ,  2.        , 10.        ,  3.        ,
        0.13863969])

In [37]:
df['babies'].unique()

array([ 0,  1,  2, 10,  9])

In [38]:
df['meal'].unique()

array(['BB', 'FB', 'HB', 'SC', 'Undefined'], dtype=object)

In [39]:
df['country'].unique()

array(['PRT', 'GBR', 'USA', 'ESP', 'IRL', 'FRA', 'others', 'ROU', 'NOR',
       'OMN', 'ARG', 'POL', 'DEU', 'BEL', 'CHE', 'CN', 'GRC', 'ITA',
       'NLD', 'DNK', 'RUS', 'SWE', 'AUS', 'EST', 'CZE', 'BRA', 'FIN',
       'MOZ', 'BWA', 'LUX', 'SVN', 'ALB', 'IND', 'CHN', 'MEX', 'MAR',
       'UKR', 'SMR', 'LVA', 'PRI', 'SRB', 'CHL', 'AUT', 'BLR', 'LTU',
       'TUR', 'ZAF', 'AGO', 'ISR', 'CYM', 'ZMB', 'CPV', 'ZWE', 'DZA',
       'KOR', 'CRI', 'HUN', 'ARE', 'TUN', 'JAM', 'HRV', 'HKG', 'IRN',
       'GEO', 'AND', 'GIB', 'URY', 'JEY', 'CAF', 'CYP', 'COL', 'GGY',
       'KWT', 'NGA', 'MDV', 'VEN', 'SVK', 'FJI', 'KAZ', 'PAK', 'IDN',
       'LBN', 'PHL', 'SEN', 'SYC', 'AZE', 'BHR', 'NZL', 'THA', 'DOM',
       'MKD', 'MYS', 'ARM', 'JPN', 'LKA', 'CUB', 'CMR', 'BIH', 'MUS',
       'COM', 'SUR', 'UGA', 'BGR', 'CIV', 'JOR', 'SYR', 'SGP', 'BDI',
       'SAU', 'VNM', 'PLW', 'QAT', 'EGY', 'PER', 'MLT', 'MWI', 'ECU',
       'MDG', 'ISL', 'UZB', 'NPL', 'BHS', 'MAC', 'TGO', 'TWN', 'DJI',
       'STP', 'KNA

In [40]:
df['market_segment'].unique()

array(['Direct', 'Corporate', 'Online TA', 'Offline TA/TO',
       'Complementary', 'Groups', 'Undefined', 'Aviation'], dtype=object)

In [41]:
df['distribution_channel'].unique()

array(['Direct', 'Corporate', 'TA/TO', 'Undefined', 'GDS'], dtype=object)

In [42]:
df['is_repeated_guest'].unique()

array([0, 1])

In [43]:
df['previous_cancellations'].unique()

array([ 0,  1,  2,  3, 26, 25, 14,  4, 24, 19,  5, 21,  6, 13, 11])

In [44]:
df['previous_bookings_not_canceled'].unique()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 20, 21, 22, 23, 24, 25, 27, 28, 29, 30, 19, 26, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72])

In [45]:
df['reserved_room_type'].unique()

array(['C', 'A', 'D', 'E', 'G', 'F', 'H', 'L', 'B'], dtype=object)

In [46]:
df['assigned_room_type'].unique()

array(['C', 'A', 'D', 'E', 'G', 'F', 'I', 'B', 'H', 'L', 'K'],
      dtype=object)

In [47]:
df['booking_changes'].unique()

array([ 3,  4,  0,  1,  2,  5, 17,  6,  8,  7, 10, 16,  9, 13, 12, 14, 15,
       18, 11])

In [48]:
df['deposit_type'].unique()

array(['No Deposit', 'Refundable', 'Non Refund'], dtype=object)

In [49]:
df['agent'].unique()

array([  0., 304., 240., 303.,  15., 241.,   8., 250., 115.,   5., 175.,
       134., 156., 243., 242.,   3., 105.,  40., 147., 306., 184.,  96.,
         2., 127.,  95., 146.,   9., 177.,   6., 143., 244., 149., 167.,
       300., 171., 305.,  67., 196., 152., 142., 261., 104.,  36.,  26.,
        29., 258., 110.,  71., 181.,  88., 251., 275.,  69., 248., 208.,
       256., 314., 126., 281., 273., 253., 185., 330., 334., 328., 326.,
       321., 324., 313.,  38., 155.,  68., 335., 308., 332.,  94., 348.,
       310., 339., 375.,  66., 327., 387., 298.,  91., 245., 385., 257.,
       393., 168., 405., 249., 315.,  75., 128., 307.,  11., 436.,   1.,
       201., 183., 223., 368., 336., 291., 464., 411., 481.,  10., 154.,
       468., 410., 390., 440., 495., 492., 493., 434.,  57., 531., 420.,
       483., 526., 472., 429.,  16., 446.,  34.,  78., 139., 252., 270.,
        47., 114., 301., 193., 182., 135., 350., 195., 352., 355., 159.,
       363., 384., 360., 331., 367.,  64., 406., 16

In [50]:
df['company'].unique()

array([  0., 110., 113., 270., 178., 240., 154., 144., 307., 268.,  59.,
       204., 312., 318.,  94., 274., 174., 195., 223., 317., 281., 118.,
        53., 286.,  12.,  47., 324., 342., 373., 371.,  86.,  82., 218.,
        88.,  31., 397., 392., 405., 331., 367.,  20.,  83., 416.,  51.,
       395., 102.,  34.,  84., 360., 394., 457., 382., 461., 478., 112.,
       486., 421.,   9., 308., 135., 224., 504., 269., 356., 498., 390.,
       513., 203., 263., 477., 521., 169., 515., 445., 337., 251., 428.,
       292., 388., 130., 250., 355., 254., 543., 531., 528.,  62., 120.,
        42.,  81., 116., 530., 103.,  39.,  16.,  92.,  61., 501., 165.,
       291., 290.,  43., 325., 192., 108., 200., 465., 287., 297., 490.,
       482., 207., 282., 437., 225., 329., 272.,  28.,  77., 338.,  72.,
       246., 319., 146., 159., 380., 323., 511., 407., 278.,  80., 403.,
       399.,  14., 137., 343., 346., 347., 349., 289., 351., 353.,  54.,
        99., 358., 361., 362., 366., 372., 365., 27

In [51]:
df['days_in_waiting_list'].unique()

array([  0,  50,  47,  65, 122,  75, 101, 150, 125,  14,  60,  34, 100,
        22, 121,  61,  39,   5,   1,   8, 107,  43,  52,   2,  11, 142,
       116,  13,  44,  97,  83,   4, 113,  18,  20, 185,  93, 109,   6,
        37, 105, 154,  64,  99,  38,  48,  33,  77,  21,  80,  59,  40,
        58,  89,  53,  49,  69,  87,  91,  57, 111,  79,  98,  85,  63,
        15,   3,  41, 224,  31,  56, 187, 176,  71,  55,  96, 236, 259,
       207, 215, 160, 120,  30,  32,  27,  62,  24, 108, 147, 379,  70,
        35, 178, 330, 223, 174, 162, 391,  68, 193,  10,  76,  16,  28,
         9, 165,  17,  25,  46,   7,  84, 175, 183,  23, 117,  12,  54,
        26,  73,  45,  19,  42,  72,  81,  74, 167,  36])

In [52]:
df['customer_type'].unique()

array(['Transient', 'Contract', 'Transient-Party', 'Group'], dtype=object)

In [53]:
df['adr'].unique()

array([  0.  ,  75.  ,  98.  , ..., 266.75, 209.25, 157.71])

In [54]:
df['required_car_parking_spaces'].unique()

array([0, 1, 2, 8, 3])

In [55]:
df['total_of_special_requests'].unique()

array([0, 1, 3, 2, 4, 5])

In [56]:
df['reservation_status'].unique()

array(['Check-Out', 'Canceled', 'No-Show'], dtype=object)

In [57]:
df['reservation_status_date'].unique()

array(['2015-07-01', '2015-07-02', '2015-07-03', '2015-05-06',
       '2015-04-22', '2015-06-23', '2015-07-05', '2015-07-06',
       '2015-07-07', '2015-07-08', '2015-05-11', '2015-07-15',
       '2015-07-16', '2015-05-29', '2015-05-19', '2015-06-19',
       '2015-05-23', '2015-05-18', '2015-07-09', '2015-06-02',
       '2015-07-13', '2015-07-04', '2015-06-29', '2015-06-16',
       '2015-06-18', '2015-06-12', '2015-06-09', '2015-05-26',
       '2015-07-11', '2015-07-12', '2015-07-17', '2015-04-15',
       '2015-05-13', '2015-07-10', '2015-05-20', '2015-05-12',
       '2015-07-14', '2015-06-17', '2015-05-01', '2015-03-30',
       '2015-07-19', '2015-06-03', '2015-06-26', '2015-05-14',
       '2015-07-20', '2015-05-07', '2015-05-28', '2015-04-13',
       '2015-03-25', '2015-07-21', '2015-06-27', '2015-07-18',
       '2015-07-23', '2015-06-08', '2015-06-22', '2015-06-24',
       '2015-03-05', '2015-06-01', '2015-04-24', '2015-07-22',
       '2015-05-27', '2015-04-06', '2015-04-11', '2015-

## ***Understanding The Hotel Booking Data***

In [58]:
df.groupby('hotel')['is_canceled'].value_counts()

hotel         is_canceled
City Hotel    0              37239
              1              16035
Resort Hotel  0              25982
              1               7974
Name: is_canceled, dtype: int64

**OBSERVATION:**

* Number of booking in city hotel is more compare to resort hotel.
* Cancelations in Resort hotel is less compared to city hotel.

In [59]:
weekdays = df.stays_in_week_nights.value_counts().sum()
weekends = df.stays_in_weekend_nights.value_counts().sum()
print(f"Bookings on weekdays are {weekdays} \nBookings on weekends are {weekends}")

Bookings on weekdays are 87230 
Bookings on weekends are 87230


**OBSERVATION**

Number of bookings on weekdays and weekends are same.

In [60]:
df.arrival_date_month.value_counts(normalize=True)

August       0.128878
July         0.115132
May          0.095655
April        0.090565
June         0.088914
March        0.085853
October      0.079342
September    0.076602
February     0.069735
December     0.058604
November     0.057010
January      0.053709
Name: arrival_date_month, dtype: float64

**OBSERVATION**

August is the most occupied (busiest) month with 0.128879 bookings and January is the most unoccupied month with 0.053709 bookings.

In [61]:
df.market_segment.value_counts(normalize=True)

Online TA        0.591001
Offline TA/TO    0.158833
Direct           0.135045
Groups           0.056426
Corporate        0.048149
Complementary    0.007933
Aviation         0.002591
Undefined        0.000023
Name: market_segment, dtype: float64

**OBSERVATION**

Around 47% of bookings are made via Online Travel Agents, almost 20% of bookings are made via Offline Travel Agents and less than 20% are Direct bookings without any other agents.

In [62]:
df.arrival_date_year.value_counts(normalize=True)

2016    0.485074
2017    0.362639
2015    0.152287
Name: arrival_date_year, dtype: float64

**OBSERVATION**

47% bookings were done in 2016, 34% in 2017 and 18 percent in 2015.. We can see increasing tendency in bookings year wise.

In [63]:
df.meal.value_counts(normalize=True)

BB           0.778482
SC           0.107658
HB           0.104093
Undefined    0.005640
FB           0.004127
Name: meal, dtype: float64

**OBSERVATION**

Out of the meals, BB (Bed & Breakfast) is the most ordered meal which is around 77.2%, followed by HB(Half Board), SC(no meal package), Undefined and FB (Full Board).

In [64]:
df.customer_type.value_counts(normalize=True)

Transient          0.823822
Transient-Party    0.134036
Contract           0.035939
Group              0.006202
Name: customer_type, dtype: float64

**OBSERVATION**

Transient type of customers are the more around 75%.

In [65]:
df.reservation_status.value_counts(normalize=True)

Check-Out    0.724762
Canceled     0.263625
No-Show      0.011613
Name: reservation_status, dtype: float64

**OBSERVATION**

We can see 63% visitors checked out and 36% bookings were canceled